src/transformers/modeling_gpt2.py

```diff
@@ -701,10 +701,23 @@ def prepare_inputs_for_generation(self, input_ids, past=None, **kwargs):
         if past:
             input_ids = input_ids[:, -1].unsqueeze(-1)
 
+        attention_mask = kwargs.get("attention_mask", None)
+        position_ids = kwargs.get("position_ids", None)
+
+        if attention_mask is not None and position_ids is None:
+            # create postion_ids on the fly for batch generation
+            position_ids = attention_mask.long().cumsum(-1) - 1
+            position_ids.masked_fill_(attention_mask == 0, 1)
+            if past:
+                position_ids = position_ids[:, -1].unsqueeze(-1)
+        else:
+            position_ids = None
```

Comment:
@patrickvonplaten
Now that you add 
`position_ids = kwargs.get("position_ids", None)`
I think we can get rid of 
`        else:
            position_ids = None
`

Also inspired by this related PR #7355, I think we should move all the `if past`  together, just above `return` 

Should I add another commit?

Reply:
No strong opinions on this, will let @patrickvonplaten decide to merge with or without this

Reply:
@cccntu - yeah I thought about this as well. The problem with this and PR #7355 and passing `position_ids` is that we would have to incrementally add new tokens to position_ids in `generate()` which would be pretty hacky since not all models support `position_ids` => so I'd rather not do this before doing a bigger refactor of generate, see: https://github.com/huggingface/transformers/pull/6949 (will continue on the bigger refactor soon).

 We can always change that later without breaking backwards compatibility.